### PR TITLE
Kyuubi Chat Engine supports Chinese questions and HTTP proxy

### PR DESCRIPTION
**Pull request by cxzl25**
### _Why are the changes needed?_

- Support Chinese question
- Support proxy settings
- Support setting timeout

image

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 1]
```diff
@@ -2667,6 +2667,20 @@ object KyuubiConf {
...
+  val ENGINE_CHAT_PROXY_HTTP_URL: OptionalConfigEntry[String] =
+    buildConf("kyuubi.engine.chat.proxy.url")
+      .doc("Whether to use a proxy to access Chat API. e.g. http://127.0.0.1:1087")
+      .version("1.8.0")
+      .stringConf
+      .createOptional
+
+  val ENGINE_CHAT_HTTP_TIMEOUT: ConfigEntry[Long] =
+    buildConf("kyuubi.engine.chat.http.timeout")
+      .doc("Timeout for calling Chat API")
```
Comment (bowenliang123) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2679]:
```suggestion
      .doc("Timeout in seconds of API calling in Chat engine")
```

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -2667,6 +2667,20 @@ object KyuubiConf {
       .stringConf
       .createOptional
 
+  val ENGINE_CHAT_PROXY_HTTP_URL: OptionalConfigEntry[String] =
+    buildConf("kyuubi.engine.chat.proxy.url")
+      .doc("Whether to use a proxy to access Chat API. e.g. http://127.0.0.1:1087")
```
Comment (bowenliang123) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2672]:
```suggestion
      .doc("HTTP proxy url for API calling in Chat engine. e.g. http://127.0.0.1:1087")
```

**Comment by bowenliang123** [comment]
image
HTTP proxy and Chinese question work fine.

**Review by bowenliang123 — APPROVED** [review]
LGTM. Great work.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 3]
```diff
@@ -2667,6 +2667,20 @@ object KyuubiConf {
       .stringConf
       .createOptional
 
+  val ENGINE_CHAT_PROXY_HTTP_URL: OptionalConfigEntry[String] =
+    buildConf("kyuubi.engine.chat.proxy.url")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2671]: let's use `kyuubi.engine.chat.gpt.` namespace

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2671]: `kyuubi.engine.chat.gpt.http.proxy`

Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2671]: And maybe we can support env `http_proxy` `HTTP_PROXY` in the future

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 1 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:2679]: maybe we can respect the http client parameter here

File: externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala [hunk 4]
```diff
@@ -32,9 +33,21 @@ import org.apache.kyuubi.engine.chat.provider.ChatProvider.mapper
 
 class ChatGPTProvider(conf: KyuubiConf) extends ChatProvider {
 
-  val token = conf.get(KyuubiConf.ENGINE_CHAT_GPT_API_KEY).get
+  private val token = conf.get(KyuubiConf.ENGINE_CHAT_GPT_API_KEY).get
```
Comment (pan3793) [inline externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala:36]: TODO(in future PR): improvement the error message if user does not provide the api key.

File: externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala [hunk 5]
```diff
@@ -32,9 +33,25 @@ import org.apache.kyuubi.engine.chat.provider.ChatProvider.mapper
...
-  val httpClient: CloseableHttpClient = HttpClientBuilder.create().build()
+  private val httpClient: CloseableHttpClient = HttpClientBuilder.create().build()
+
+  private val requestConfig = {
+    val connectTimeout = conf.get(KyuubiConf.ENGINE_CHAT_GPT_HTTP_CONNECT_TIMEOUT).asInstanceOf[Int]
+    val socketTimeout = conf.get(KyuubiConf.ENGINE_CHAT_GPT_HTTP_SOCKET_TIMEOUT).asInstanceOf[Int]
+    val builder: RequestConfig.Builder = RequestConfig.custom()
+      .setConnectionRequestTimeout(connectTimeout)
+      .setConnectTimeout(connectTimeout)
+      .setSocketTimeout(socketTimeout)
```
Comment (pan3793) [inline externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala:49]:
```suggestion
      .setConnectTimeout(connectTimeout)
      .setSocketTimeout(socketTimeout)
```

File: externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala [hunk 6]
```diff
@@ -32,9 +33,24 @@ import org.apache.kyuubi.engine.chat.provider.ChatProvider.mapper
 
 class ChatGPTProvider(conf: KyuubiConf) extends ChatProvider {
 
-  val token = conf.get(KyuubiConf.ENGINE_CHAT_GPT_API_KEY).get
+  private val gptApiKey = conf.get(KyuubiConf.ENGINE_CHAT_GPT_API_KEY).getOrElse {
+    throw new IllegalArgumentException(
+      s"${KyuubiConf.ENGINE_CHAT_GPT_API_KEY.key} must be configured")
```
Comment (pan3793) [inline externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala:38]:
```suggestion
      s"'${KyuubiConf.ENGINE_CHAT_GPT_API_KEY.key}' must be configured, " +
        s"which could be got at https://platform.openai.com/account/api-keys")
```

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to master

**Pull request closed without merging**